snuba/consumers/consumer.py

```diff
@@ -62,6 +67,95 @@ def __reduce_ex__(
             return type(self), (self.rows, self.origin_timestamp)
 
 
+class ValuesInsertBatch(NamedTuple):
+    rows: Sequence[bytes]
+    origin_timestamp: Optional[datetime]
+
+    @staticmethod
+    def encode_value(value: Any) -> str:
+        if isinstance(value, str):
+            return value
+
+        if isinstance(value, int):
+            return str(value)
+
+        if isinstance(value, list):
+            for x in value:
+                if not isinstance(x, int):
+                    raise TypeError("ints should only contain arrays")
+
+            return "[" + ",".join([str(x) for x in value]) + "]"
+
+        return "0"
+
+    @staticmethod
+    def encode_row(row: WriterTableRow) -> bytes:
+        ordered_columns = [
+            ValuesInsertBatch.encode_value(row[column])
+            for column in DISTRIBUTIONS_COLUMN_SET
+        ]
+        ordered_columns_str = ",".join(ordered_columns)
+        return f"({ordered_columns_str})".encode("utf-8")
+
+
+class ValuesInsertBatchWriter(ProcessingStep[ValuesInsertBatch]):
+    def __init__(self, writer: BatchWriter[bytes], metrics: MetricsBackend,) -> None:
+        self.__writer = writer
+        self.__metrics = metrics
+        self.__messages: MutableSequence[Message[ValuesInsertBatch]] = []
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Message[ValuesInsertBatch]) -> None:
+        self.__messages.append(message)
+
+    def close(self) -> None:
+        logger.error(self.__messages)
+        message = self.__messages[0]
+        logger.error(message.payload.rows)
+
+        self.__closed = True
+
+        if not self.__messages:
+            return
+
+        write_start = time.time()
+        self.__writer.write(
+            itertools.chain.from_iterable(
+                message.payload.rows for message in self.__messages
+            )
+        )
+        write_finish = time.time()
+
+        for message in self.__messages:
+            self.__metrics.timing(
+                "latency_ms", (write_finish - message.timestamp.timestamp()) * 1000
+            )
+            if message.payload.origin_timestamp is not None:
+                self.__metrics.timing(
+                    "end_to_end_latency_ms",
+                    (write_finish - message.payload.origin_timestamp.timestamp())
+                    * 1000,
+                )
+        self.__metrics.timing("batch_write_ms", write_finish - write_start)
```

Comment:
It would be a good idea to differentiate the name of the metrics for value based writes v/s JSON writes. That way you could compare the two if you wanted to.